### PR TITLE
Make lookup data a separate crate

### LINKED ISSUE #2050: Inconsistent view of in-memory lookup data while a Wasm function executes
Because of the way the Oak Functions loader refreshes data in the background, it might happen that if a Wasm module performs multiple lookups, a new version of the lookup data might become available between lookups from the same invocation, which may result in inconsistent behaviour of the module.

We could solve this by acquiring a Read lock on the data at invocation time (or the first time the module invokes the lookup ABI function), and hold it until the invocation is terminated (instead of acquiring it and releasing it as part of each lookup ABI call).

When the loader needs to update the data from the background process, it would still acquire a write lock for it, which would be guaranteed to be exclusive. This is fine because this write lock would be held for a very short time (just to swap the reference behind it), so there should not be too much contention.

I am not sure what priority the read and write portion of the RWLock would get though, and whether it is possible that the background process would starve if there is a constant stream of requests that keep requesting the read lock, or if the write lock has higher priority (and with what characteristics).

### PR DESCRIPTION
**Pull request by conradgrobler**
The change splits the lookup data functionality into two components:

- A `LookupDataManager` that owns the in-memory lookup data. It is moved into a separate crate, and will be made `no_std`-compatible in future, and will remain in the TCB
- A `LookupDataRefresher` that refreshes the lookup data owned by the manager from a file or URL, and can be moved outside the TCB in  future

The change also upgrades the `regex` crate to v1.5.5 to resolve a cargo-deny advisory.

Fixes #2050

### CODE REVIEW

File: oak_functions/loader/src/lookup.rs [hunk 1]
```diff
@@ -77,43 +64,35 @@ pub fn storage_get_item(
...
-                .logger
-                .log_sensitive(Level::Debug, "storage_get_item(): value not found");
+            extension.log_debug("storage_get_item(): value not found");
             Err(OakStatus::ErrStorageItemNotFound)
         }
     }
 }
 
-impl OakApiNativeExtension for LookupExtension {
+impl OakApiNativeExtension for LookupData<Logger> {
```
Comment (tiziano88) [inline oak_functions/loader/src/lookup.rs:95]: This should be implemented for any Logger, right? (not that it matters for us, but it seems too specific at the moment)

**Review by tiziano88 — APPROVED** [review]
I'm still not sure why we need a lookup manager *and* a refresher, but perhaps it will be clearer (to me) once they are in fact on different sides of the trust boundary.

File: oak_functions/loader/src/lookup.rs (hunk 1 above)
Comment (conradgrobler) [reply oak_functions/loader/src/lookup.rs:95]: Done.

File: oak_functions/lookup/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    /// Creates an instance of LookupData populated with the given entries.
+    pub fn for_test(entries: Data, logger: L) -> Self {
+        let data = RwLock::new(Arc::new(entries));
+        Self { data, logger }
+    }
+
+    /// Updates the backing data that will be used by new `LookupData` instances.
+    pub fn update_data(&self, data: Data) {
+        *self.data.write().expect("could not lock data for read") = Arc::new(data);
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak_functions/lookup/src/lib.rs:65]: Done.

File: oak_functions/lookup/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    /// Whether the backing data is empty.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    /// Logs an error message.
+    ///
+    /// The code assumes the message might contain sensitive information.
+    pub fn log_error(&self, message: &str) {
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak_functions/lookup/src/lib.rs:111]: Yes, these are for convenience. These will likely be removed once #2576 is done and there is no longer code in a different crate using the logger instance on the lookup data extension.

**Review by conradgrobler — COMMENTED** [review]
> I'm still not sure why we need a lookup manager and a refresher, but perhaps it will be clearer (to me) once they are in fact on different sides of the trust boundary.

This will be somewhat improved once #2576 is completed. The LookupFactory and LookupDataManger will be merged into one struct. I agree that the distinction will be clearer once the refresher is moved outside of the TCB.

**Pull request merged**